### ISSUE TITLE
Reading and writing part 3: web applications

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8464]
Reading and writing, part 3: web applications<br>
<br>
https://ift.tt/87suaHU<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<div>
<p>This article is part of a five-part series on race conditions. The complete series is:</p>
<p>&ndash;</p>
<p>In <a href="https://blog.jcoglan.com/2020/09/28/reading-and-writing-part-1/">part 1</a> of this series, we met the idea of the <code>Location</code> type, a resource that can be read from and written to without co-ordination, and which therefore allows race conditions to happen. In <a href="https://blog.jcoglan.com/2020/10/05/reading-and-writing-part-2/">part 2</a> we saw how this pattern occurs when interacting with files and databases, and we learned many ways that databases let us address its problems using <em>pessimistic</em> and <em>optimistic</em> control techniques. In this article, we&rsquo;ll put this in context by looking at how read and write patterns in web applications can lead to race conditions and inconsistent data.</p>
<p>To begin with, let&rsquo;s look at probably the most common form of race condition in such applications: uniqueness violations. It&rsquo;s common for applications to require certain fields, like a user&rsquo;s email address, to be unique, and Rails has historically not dealt with this requirement robustly. <a href="http://www.bailis.org/papers/feral-sigmod2015.pdf">Feral Concurrency Control</a> gives a good account of this problem, including a reminder that, at least in its early days, the design approach in Rails was that all logic should reside in application code, not in the database. The database should be treated as a simple data store for reading and writing records, and nothing more &ndash; that is, it should be thought of as a <code>Location</code>.</p>
<blockquote>
<p>I don&rsquo;t want my database to be clever! &hellip; I consider stored procedures and constraints vile and reckless destroyers of coherence. No, Mr. Database, you can not have my business logic. Your procedural ambitions will bear no fruit and you&rsquo;ll have to pry that logic from my dead, cold object-oriented hands&hellip; I want a single layer of cleverness: My domain model.</p>
<p><cite>David Heinemeier Hansson</cite></p>
</blockquote>
<p>The problem with this assertion is that once you&rsquo;re running more than one instance of your application &ndash; whether that&rsquo;s using multiple threads, processes, or server machines &ndash; you no longer have a <em>single</em> &ldquo;layer of cleverness&rdquo;, you have multiple clients talking concurrently to your database. In that environment, many of the interfaces Rails provides for guaranteeing consistency aren&rsquo;t effective unless they&rsquo;re combined with database concurrency control.</p>
<p>I should emphasize that the problems I am about to demonstrate are not unique to Rails, they&rsquo;re commonplace in web application logic across many platforms. Rails is simply a well-known example, with which I&rsquo;m most familiar. It has also had an enormous influence on how web apps in general are written; tools like Rails are how many people have <em>learned</em> to develop applications, and so its lack of decent concurrency control contributes to many people never learning how to notice, diagnose or fix the bugs it leads to. Tools aren&rsquo;t just a means of getting work done, they&rsquo;re a means of <em>understanding</em> the work itself, and the ergonomic effect of a tool making it easy to write certain categories of bug has an impact beyond the use of that tool in particular.</p>
<p>Here&rsquo;s an example of how the &ldquo;Rails Way&rdquo; for ensuring uniqueness can be easily broken. Say we have a table that stores <em>settings</em>, that is configuration data, and it should be treated as a key-value store. Many open-source applications include such a table, so that end users can tweak the system&rsquo;s functionality. In Rails we would declare a schema for this table with two columns, <code>name</code> and <code>value</code>, and a model class to represent the table.</p>
<div>
<div>
<pre><code><span>ActiveRecord</span><span>::</span><span>Schema</span><span>.</span><span>define</span> <span>do</span>
  <span>create_table</span> <span>:settings</span><span>,</span> <span>force: </span><span>true</span> <span>do</span> <span>|</span><span>t</span><span>|</span>
    <span>t</span><span>.</span><span>string</span> <span>:name</span>
    <span>t</span><span>.</span><span>string</span> <span>:value</span>
  <span>end</span>
<span>end</span>

<span>class</span> <span>Setting</span> <span>&lt;</span> <span>ActiveRecord</span><span>::</span><span>Base</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>Now, we&rsquo;d like to design an API for assigning settings in this table, making sure each <code>name</code> appears only once and thus corresponds to a unique value. Here&rsquo;s the experiment we&rsquo;ll run to determine whether each implementation is correct. We&rsquo;ll start a set of threads, each of which will try to assign a different value for the <code>timezone</code> setting. Once the threads have completed, we&rsquo;ll count the number of rows in the <code>settings</code> table.</p>
<div>
<div>
<pre><code><span>threads</span> <span>=</span> <span>(</span><span>1</span><span>..</span><span>4</span><span>).</span><span>map</span> <span>do</span> <span>|</span><span>n</span><span>|</span>
  <span>Thread</span><span>.</span><span>new</span> <span>do</span>
    <span>Setting</span><span>.</span><span>set_value</span><span>(</span><span>"timezone"</span><span>,</span> <span>"UTC+</span><span>#{</span><span>n</span><span>}</span><span>"</span><span>)</span>
  <span>end</span>
<span>end</span>

<span>threads</span><span>.</span><span>each</span><span>(</span><span>&amp;</span><span>:join</span><span>)</span>
<span>p</span> <span>[</span><span>:count</span><span>,</span> <span>Setting</span><span>.</span><span>count</span><span>]</span>
</code></pre>
</div>
</div>
<p>As we&rsquo;ve only assigned to a single setting name, we expect only one row in the table. As I mentioned in a previous article, I am using <a href="https://www.jruby.org/">JRuby</a> here as it supports truly parallel thread execution.</p>
<p>Historically, by far the most common way to see this implemented is to use <code>find_or_create_by</code> to check if a record exists with certain properties, and create one if not. Having ensured the existence of a <code>Setting</code> with the right <code>name</code>, we can assign a <code>value</code> to it.</p>
<div>
<div>
<pre><code><span>class</span> <span>Setting</span>
  <span>def</span> <span>self</span><span>.</span><span>set_value</span><span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span>
    <span>setting</span> <span>=</span> <span>find_or_create_by</span><span>(</span><span>name: </span><span>name</span><span>)</span>
    <span>setting</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>value</span><span>)</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>When I run this implementation, there are almost always four records in <code>settings</code>, all with the same <code>name</code>, so this is clearly not workable. Let&rsquo;s look at the SQL this generates to understand what&rsquo;s happening:</p>
<div>
<div>
<pre><code><span>SELECT</span> <span>*</span> <span>FROM</span> <span>settings</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>BEGIN</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>)</span> <span>RETURNING</span> <span>id</span><span>;</span>
<span>COMMIT</span><span>;</span>
<span>BEGIN</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+1'</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>It begins by doing a <code>SELECT</code> to see if a record with <code>name = 'timezone'</code> exists. If this returns nothing, then it performs an <code>INSERT</code> for that <code>name</code> inside a transaction; this will return the <code>id</code> of the new record. Given that <code>id</code>, it then issues an <code>UPDATE</code> to set the <code>value</code> on the new record, in a separate transaction. If the initial <code>SELECT</code> does return a record, then the middle transaction containing the <code>INSERT</code> does not happen:</p>
<div>
<div>
<pre><code><span>SELECT</span> <span>*</span> <span>FROM</span> <span>settings</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>BEGIN</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+2'</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>Note if there is already a transaction in progress, then Rails does not insert transactions around the <code>INSERT</code> or <code>UPDATE</code> queries &ndash; it relies on the existing transaction to make any associated life-cycle callbacks atomic.</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>...</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>settings</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>)</span> <span>RETURNING</span> <span>id</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+1'</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>...</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>We can see why this doesn&rsquo;t work: we&rsquo;re treating the <code>settings</code> table as a <code>Location</code> &ndash; not an individual record in the table, but the table itself. We read from the table to check if any records with a certain <code>name</code> exist, and then based on the result of that we write to the table to add an additional row, which always succeeds. Since this operation can be characterised by the <code>Location</code> type, it is inherently prone to race conditions. In this case, it means two tasks can both run the first <code>SELECT</code>, both see that no matching records exist, and therefore both <code>INSERT</code> one.</p>
<div>
<div>
<pre><code><span>Task</span> <span>1</span>                          <span>|</span> <span>Task</span> <span>2</span>
<span>--------------------------------+--------------------------------</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>settings</span> <span>...</span>      <span>|</span>
    <span>--&gt; no results |</span>
                                <span>|</span> <span>SELECT</span> <span>*</span> <span>FROM</span> <span>settings</span> <span>...</span>
                                <span>|</span>     <span>--&gt; no results</span>
<span>BEGIN</span>                           <span>|</span>
                                <span>|</span> <span>BEGIN</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>...</span>        <span>|</span>
                                <span>|</span> <span>INSERT</span> <span>INTO</span> <span>settings</span> <span>...</span>
<span>COMMIT</span>                          <span>|</span>
                                <span>|</span> <span>COMMIT</span>
</code></pre>
</div>
</div>
<p>Putting the <code>INSERT</code> in a transaction here does not give us any protection from the database &ndash; these <code>INSERT</code> queries are both valid, and don&rsquo;t conflict with each other since the database doesn&rsquo;t know <code>name</code> needs to be unique. So, all the above queries succeed and two <code>INSERT</code> queries are done with the same <code>name</code>.</p>
<p>This race condition is an example of <em>write skew</em>: two transactions read from the same set of records, and then both write something that would be valid on its own, but invalidates a decision made by the other transaction. In this case, both transactions decide to create a record with <code>name = 'timezone'</code> because they believe that none exists, but when both are committed then our requirement that <code>name</code> is unique has been violated. It&rsquo;s a race condition because if these tasks were executed sequentially, this mistake wouldn&rsquo;t happen: the later task would see the record insert by the earlier one, and decide to do nothing.</p>
<p>A <em>lost update</em> can be viewed as a special case of write skew in which both writes go to the same location, and so one overwrites the other. In general, write skews involve both writes being kept and together creating an invalid state. When we examined lost updates with the <code>counters</code> example, we saw that in PostgreSQL the <em>repeatable read</em> isolation level can detect lost updates and reject the transactions that cause them. <em>Some</em> write skew scenarios can be detected this way, but not all, because in general the writes affect different records.</p>
<p>They also can&rsquo;t be addressed by locking. For example, let&rsquo;s try using the <code>lock</code> method inside a transaction and look at the generated SQL:</p>
<div>
<div>
<pre><code><span>class</span> <span>Setting</span>
  <span>def</span> <span>self</span><span>.</span><span>set_value</span><span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span>
    <span>transaction</span> <span>do</span>
      <span>setting</span> <span>=</span> <span>lock</span><span>.</span><span>find_or_create_by</span><span>(</span><span>name: </span><span>name</span><span>)</span>
      <span>setting</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>value</span><span>)</span>
    <span>end</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>As expected, this generates a <code>FOR UPDATE</code> clause on the initial <code>SELECT</code>. However, it does not prevent the race from happening, because <code>FOR UPDATE</code> locks the records that happen to match the <code>WHERE</code> clause <em>right now</em>, it does not prevent new records matching the clause being created, existing records being deleted, or existing records being modified to match.</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>settings</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span> <span>LIMIT</span> <span>1</span> <span>FOR</span> <span>UPDATE</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>)</span> <span>RETURNING</span> <span>id</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+1'</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>That is, it locks specific records, rather than the abstract set of { all records where <code>name = 'timezone'</code> }. Therefore, two transactions can both run the <code>INSERT</code> query and create a duplicate record in the process.</p>
<p>The same approach is used in the <a href="https://api.rubyonrails.org/classes/ActiveRecord/Validations/ClassMethods.html#method-i-validates_uniqueness_of"><code>validates_uniqueness_of</code></a> macro, and its own documentation will tell you it&rsquo;s not safe under concurrent requests, as we&rsquo;ve just seen. They recommend instead that a unique index is added to the database, which will cause it to reject any <code>INSERT</code> or <code>UPDATE</code> that creates duplicate values on the given columns. This sounds promising; as we&rsquo;ve seen, simply making <code>write()</code> fallible makes it possible to avoid some consistency bugs.</p>
<div>
<div>
<pre><code><span>ActiveRecord</span><span>::</span><span>Schema</span><span>.</span><span>define</span> <span>do</span>
  <span>create_table</span> <span>:settings</span><span>,</span> <span>force: </span><span>true</span> <span>do</span> <span>|</span><span>t</span><span>|</span>
    <span>t</span><span>.</span><span>string</span> <span>:name</span>
    <span>t</span><span>.</span><span>string</span> <span>:value</span>

    <span>t</span><span>.</span><span>index</span> <span>[</span><span>:name</span><span>],</span> <span>unique: </span><span>true</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>We can then reverse the order of operations: first try an <code>INSERT</code>, and if that fails, detect the <code>RecordNotUnique</code> exception and <code>UPDATE</code> the existing record. This exception type was added in Rails 3.0, released in 2010. Unfortunately, documentation for using it correctly is scant, and it&rsquo;s easily misused. Here&rsquo;s a first attempt:</p>
<div>
<div>
<pre><code><span>class</span> <span>Setting</span>
  <span>def</span> <span>self</span><span>.</span><span>set_value</span><span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span>
    <span>create</span><span>(</span><span>name: </span><span>name</span><span>,</span> <span>value: </span><span>value</span><span>)</span>
  <span>rescue</span> <span>ActiveRecord</span><span>::</span><span>RecordNotUnique</span>
    <span>where</span><span>(</span><span>name: </span><span>name</span><span>).</span><span>update_all</span><span>(</span><span>value: </span><span>value</span><span>)</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>When a matching record does not exist, then we perform the <code>INSERT</code> with all the required values, and we&rsquo;re done:</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>,</span> <span>'UTC+1'</span><span>);</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>If a matching record does exist, the <code>INSERT</code> fails. The transaction that was started inside the Rails <code>create</code> method is rolled back, and then we run an <code>UPDATE</code> to set the desired <code>value</code> on the existing record.</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>,</span> <span>'UTC+2'</span><span>);</span>
<span>ROLLBACK</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+2'</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span><span>;</span>
</code></pre>
</div>
</div>
<p>This will often work fine, although the control flow is a little awkward outside of trivial cases because of the exception handling. The deeper problem is that this doesn&rsquo;t work if a transaction is already in progress. Most databases don&rsquo;t support nesting transactions, and as we saw above Rails won&rsquo;t initiate a transaction to create or update something if there&rsquo;s a transaction already open on the current connection. Because we catch the <code>RecordNotUnique</code> exception, and <code>create</code> does not start and manage its own transaction, no <code>ROLLBACK</code> happens:</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span> <span>-- existing transaction started outside of Setting.set_value</span>
<span>...</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>,</span> <span>'UTC+2'</span><span>);</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+2'</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span><span>;</span>
<span>...</span>
</code></pre>
</div>
</div>
<p>Because the <code>INSERT</code> was rejected, PostgreSQL now considers this transaction failed; it&rsquo;s not aborted, but running further queries in it is now an error. But, we don&rsquo;t want to <code>ROLLBACK</code>, as that would undo whatever changes were already done before the <code>INSERT</code> happened &ndash; we just want to recover from the failed <code>INSERT</code>, not throw the whole transaction away.</p>
<p>On MySQL, the effect is even worse. Because of the way it <a href="https://dev.mysql.com/doc/refman/8.0/en/innodb-locks-set.html">locks index records</a> during <code>INSERT</code> and <code>UPDATE</code> operations, two transactions executing the above queries can <a href="https://en.wikipedia.org/wiki/Deadlock">deadlock</a> each other. The failed <code>INSERT</code> does not render the transaction unusable, and when a duplicate-key error happens, <code>INSERT</code> acquires a shared lock on the index record for the given key. As the transaction is still viable, we can run an <code>UPDATE</code>, which tries to upgrade the index lock to an exclusive lock. If both transactions have executed the <code>INSERT</code> and hold a shared lock, neither can acquire the exclusive lock until the other commits or aborts.</p>
<p>This is problematic because your code will usually appear to work fine during development, unless you actively check for concurrency bugs, but as soon as you put it in production you&rsquo;ll get errors from users deadlocking the database. By eagerly invalidating the transaction after the failed <code>INSERT</code>, PostgreSQL forces you to write something that will work predictably in production. An important element of ergonomic design in software systems is the decision to eagerly report errors for things that aren&rsquo;t guaranteed to work, rather than letting the code work sometimes and letting you go to production before finding out it might fail. In this case, PostgreSQL protects you from making this mistake, but as we&rsquo;ve seen there are other cases where it won&rsquo;t unless you explicitly ask it to. For example, by running transactions in the <em>read committed</em> level by default, it does not prevent lost updates or write skew.</p>
<p>To handle rolling back the failed <code>INSERT</code>, PostgreSQL, MySQL and other databases provide <a href="https://www.postgresql.org/docs/12/sql-savepoint.html">savepoints</a>. These let you set markers within a transaction and roll back to the state at a given point, without discarding the entire transaction. Rails has long had an API for savepoints, but it remains undocumented and is not widely known. We can use it in our example by calling <code>create_savepoint</code> before the <code>INSERT</code> operation, and then <code>rollback_to_savepoint</code> if it fails:</p>
<div>
<div>
<pre><code><span>class</span> <span>Setting</span>
  <span>def</span> <span>self</span><span>.</span><span>set_value</span><span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span>
    <span>sp_name</span> <span>=</span> <span>"create_setting_</span><span>#{</span><span>name</span><span>}</span><span>"</span>

    <span>transaction</span> <span>do</span>
      <span>connection</span><span>.</span><span>create_savepoint</span><span>(</span><span>sp_name</span><span>)</span>
      <span>create</span><span>(</span><span>name: </span><span>name</span><span>,</span> <span>value: </span><span>value</span><span>)</span>
    <span>rescue</span> <span>ActiveRecord</span><span>::</span><span>RecordNotUnique</span>
      <span>connection</span><span>.</span><span>rollback_to_savepoint</span><span>(</span><span>sp_name</span><span>)</span>
      <span>where</span><span>(</span><span>name: </span><span>name</span><span>).</span><span>update_all</span><span>(</span><span>value: </span><span>value</span><span>)</span>
    <span>end</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>The <code>transaction</code> call starts a new transaction if there is not one currently running; savepoints can only be used inside transactions, and this construction is safe if we call <code>set_value</code> from within another transaction. It generates the following queries when no matching records exist:</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>SAVEPOINT</span> <span>create_setting_timezone</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>,</span> <span>'UTC+1'</span><span>);</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>When a matching record does exist, a <code>ROLLBACK TO SAVEPOINT</code> happens and we can continue to use the open transaction &ndash; this also prevents deadlocks on MySQL by releasing the lock acquired by the failed <code>INSERT</code>.</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>SAVEPOINT</span> <span>create_setting_timezone</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>,</span> <span>'UTC+2'</span><span>);</span>
<span>ROLLBACK</span> <span>TO</span> <span>SAVEPOINT</span> <span>create_setting_timezone</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+2'</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>There is a more convenient way to use savepoints in Rails: the <a href="https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/DatabaseStatements.html#method-i-transaction"><code>transaction</code></a> method takes an option called <code>requires_new</code>. Databases don&rsquo;t support nested transactions, but <code>requires_new</code> lets you pretend they do, at least as far as rolling back is concerned. Normally <code>transaction</code> only starts a transaction with <code>BEGIN</code> if there isn&rsquo;t already one running. <code>requires_new</code> means that, if there&rsquo;s already a transaction running, a savepoint will be set so that any exception causes a rollback to that savepoint, rather than undoing the whole transaction. It wraps the above workflow up so we can use the same code whether we&rsquo;re already in a transaction or not.</p>
<div>
<div>
<pre><code><span>class</span> <span>Setting</span>
  <span>def</span> <span>self</span><span>.</span><span>set_value</span><span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span>
    <span>transaction</span><span>(</span><span>requires_new: </span><span>true</span><span>)</span> <span>do</span>
      <span>create</span><span>(</span><span>name: </span><span>name</span><span>,</span> <span>value: </span><span>value</span><span>)</span>
    <span>end</span>
  <span>rescue</span> <span>ActiveRecord</span><span>::</span><span>RecordNotUnique</span>
    <span>where</span><span>(</span><span>name: </span><span>name</span><span>).</span><span>update_all</span><span>(</span><span>value: </span><span>value</span><span>)</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>Note that our exception handler is <em>outside</em> the <code>transaction</code> block. To issue a <code>ROLLBACK</code>, <code>transaction</code> needs to know an exception has happened, and catching it inside the <code>transaction</code> block would prevent that. Instead, we let <code>transaction</code> roll the transaction back and then handle the exception ourselves. When there&rsquo;s no transaction in progress, this generates the following queries:</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>,</span> <span>'UTC+2'</span><span>);</span>
<span>ROLLBACK</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+2'</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span><span>;</span>
</code></pre>
</div>
</div>
<p>If we&rsquo;re inside an existing transaction, then an equivalent set of operations is done using <code>SAVEPOINT</code>, so that the whole transaction is not aborted and we can keep using it.</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>...</span>
<span>SAVEPOINT</span> <span>active_record_1</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>,</span> <span>'UTC+2'</span><span>);</span>
<span>ROLLBACK</span> <span>TO</span> <span>SAVEPOINT</span> <span>active_record_1</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+2'</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span><span>;</span>
<span>...</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>In Rails 6.0, released in 2019, a hybrid between this pattern and <code>find_or_create_by</code> was introduced, named <code>create_or_find_by</code>. It reverses the operations of <code>find_or_create_by</code> by attempting an <code>INSERT</code> first, and then doing a <code>SELECT</code> if this fails. The record returned by either of these operations can then be completed using an <code>UPDATE</code>.</p>
<div>
<div>
<pre><code><span>class</span> <span>Setting</span>
  <span>def</span> <span>self</span><span>.</span><span>set_value</span><span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span>
    <span>setting</span> <span>=</span> <span>create_or_find_by</span><span>(</span><span>name: </span><span>name</span><span>)</span>
    <span>setting</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>value</span><span>)</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p><code>create_or_find_by</code> deals with transactions correctly, starting a new one if needed and using savepoints if one is already in progress, so it prevents transactions failing on PostgreSQL or deadlocking on MySQL. When no matching record exists, we get an <code>INSERT</code> followed by an <code>UPDATE</code>:</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>)</span> <span>RETURNING</span> <span>id</span><span>;</span>
<span>COMMIT</span><span>;</span>
<span>BEGIN</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+1'</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>Or, if a transaction is already running:</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>...</span>
<span>SAVEPOINT</span> <span>active_record_1</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>)</span> <span>RETURNING</span> <span>id</span><span>;</span>
<span>RELEASE</span> <span>SAVEPOINT</span> <span>active_record_1</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+1'</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>...</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>If a matching record exists, then the <code>INSERT</code> fails, so we <code>ROLLBACK</code>, load the existing record with <code>SELECT</code> and modify it using <code>UPDATE</code>:</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>)</span> <span>RETURNING</span> <span>id</span><span>;</span>
<span>ROLLBACK</span><span>;</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>settings</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>BEGIN</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+2'</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>If there is already a transaction running, then savepoints are used, but note that the <code>INSERT</code>, <code>SELECT</code> and <code>UPDATE</code> now happen in the same transaction. If some other query deletes the matching record after the <code>INSERT</code> fails, then under <em>read committed</em> the <code>SELECT</code> will return nothing. Under <em>repeatable read</em>, the <code>SELECT</code> will return the record that caused the <code>INSERT</code> to fail, even though it no longer exists, and the <code>UPDATE</code> will trigger a serialization failure, so this can still go wrong in certain scenarios.</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>...</span>
<span>SAVEPOINT</span> <span>active_record_1</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>)</span> <span>RETURNING</span> <span>id</span><span>;</span>
<span>ROLLBACK</span> <span>TO</span> <span>SAVEPOINT</span> <span>active_record_1</span><span>;</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>settings</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+2'</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>...</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>Although <code>create_or_find_by</code> is a welcome addition in terms of producing consistent writes, it is less efficient than our <code>RecordNotUnique</code> approach above. When the <code>INSERT</code> succeeds, it only fills in the <code>name</code> column and an additional <code>UPDATE</code> is needed to add the <code>value</code>. When the <code>INSERT</code> fails, an additional <code>SELECT</code> is done to load the existing record into memory so it can be updated. This is consistent with the general Rails approach of performing updates by loading a record with <code>SELECT</code>, modifying the in-memory copy, and then writing it back with <code>UPDATE</code>, but is has some notable disadvantages.</p>
<p>In addition to requiring extra queries, <code>create_or_find_by</code> shares a flaw with <code>find_or_create_by</code>: it creates partial records. The <code>INSERT</code> only fills in the field we&rsquo;re searching on &ndash; the <code>name</code> &ndash; and leaves all other fields <code>NULL</code>. This <code>NULL</code> may never be observed, <em>if</em> <code>create_or_find_by</code> is run inside a transaction wrapping both the <code>INSERT</code> and <code>UPDATE</code> operations, and if none of your queries run in the <em>read uncommitted</em> isolation level. However, it still requires that your schema allow the other columns to be <code>NULL</code>, which may be undesirable even if it is the default for columns created by Rails.</p>
<p>Fortunately, PostgreSQL (since 9.5) and MySQL both offer a solution to all these problems: the <em>upsert</em> operation, short for <em>update or insert</em>. Rather than thinking about the exact change operations we want to make to the data, we can think instead in terms of the final state we want to produce, which is that exactly one row must exist in <code>settings</code> with the given <code>name</code>, and that row should have the given <code>value</code>. On PostgreSQL this is written as:</p>
<div>
<div>
<pre><code><span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>,</span> <span>'UTC+1'</span><span>)</span>
    <span>ON</span> <span>CONFLICT</span> <span>(</span><span>name</span><span>)</span> <span>DO</span> <span>UPDATE</span> <span>SET</span> <span>value</span><span>=</span><span>excluded</span><span>.</span><span>value</span>
    <span>RETURNING</span> <span>id</span><span>;</span>
</code></pre>
</div>
</div>
<p>On MySQL we do not have to specify which column the conflict occurs on, and the equivalent query is:</p>
<div>
<div>
<pre><code><span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>,</span> <span>'UTC+1'</span><span>)</span>
    <span>ON</span> <span>DUPLICATE</span> <span>KEY</span>
    <span>UPDATE</span> <span>name</span><span>=</span><span>VALUES</span><span>(</span><span>name</span><span>),</span> <span>value</span><span>=</span><span>VALUES</span><span>(</span><span>value</span><span>);</span>
</code></pre>
</div>
</div>
<p>This is a single atomic query that achieves the desired state without having to read anything first, and does not require a transaction. The Rails wrapper for it is called <code>upsert</code>, also added in 6.0:</p>
<div>
<div>
<pre><code><span>class</span> <span>Setting</span>
  <span>def</span> <span>self</span><span>.</span><span>set_value</span><span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span>
    <span>upsert</span><span>({</span> <span>name: </span><span>name</span><span>,</span> <span>value: </span><span>value</span> <span>},</span> <span>unique_by: </span><span>[</span><span>:name</span><span>])</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>On PostgreSQL the <code>unique_by</code> option is needed to specify the columns for the <code>ON CONFLICT</code> clause, whereas on MySQL this option must <em>not</em> be used, and will cause an error if present, so this API is not portable across databases. However, it is the first API Rails has introduced that directly maps to the functionality built into databases for precisely this scenario.</p>
<p>Not all write skew errors can be prevented with a unique index. Such indexes can prevent duplicate values from appearing in a column, but they can&rsquo;t express more complex constraints like &ldquo;there must be exactly one <code>article</code> marked as <code>featured</code> for each <code>author_id</code>&rdquo;. In general, preventing write skew requires either taking a <code>FOR UPDATE</code> lock on the existing records over which the constraint applies, or using an application-level advisory lock if the operation involves the set of matching records gaining or losing rows. For example, taking an advisory lock for the affected <code>author_id</code> would handle the aforementioned requirement.</p>
<p>The final option you have for preventing this type of bug is to run the operation in a transaction with the <em>serializable</em> isolation level. This means the database will run transactions so their result is consistent with running them in some sequential order; any operation that would lead to an ambiguous outcome due to the ordering of concurrent queries is rejected. That makes it safe to do this:</p>
<div>
<div>
<pre><code><span>BEGIN</span> <span>ISOLATION</span> <span>LEVEL</span> <span>SERIALIZABLE</span><span>;</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>settings</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'timezone'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>INSERT</span> <span>INTO</span> <span>settings</span> <span>(</span><span>name</span><span>)</span> <span>VALUES</span> <span>(</span><span>'timezone'</span><span>)</span> <span>RETURNING</span> <span>id</span><span>;</span>
<span>UPDATE</span> <span>settings</span> <span>SET</span> <span>value</span> <span>=</span> <span>'UTC+1'</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>As we saw with <code>find_or_create_by</code>, this transaction is ambiguous depending on how two transactions interleave the queries: both could run <code>SELECT</code>, get no results, and so both run the <code>INSERT</code>, whereas if they were run sequentially then only one would run the <code>INSERT</code>. That ambiguity means running these transactions concurrently under <em>serializable</em> isolation is illegal, and one of them will be rejected.</p>
<p>This means we can take our <code>find_or_create_by</code> implementation, wrap it in a serializable transaction and retry it if it&rsquo;s rejected, and it will do the right thing.</p>
<div>
<div>
<pre><code><span>class</span> <span>Setting</span>
  <span>def</span> <span>self</span><span>.</span><span>set_value</span><span>(</span><span>name</span><span>,</span> <span>value</span><span>)</span>
    <span>transaction</span><span>(</span><span>isolation: :serializable</span><span>)</span> <span>do</span>
      <span>setting</span> <span>=</span> <span>find_or_create_by</span><span>(</span><span>name: </span><span>name</span><span>)</span>
      <span>setting</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>value</span><span>)</span>
    <span>end</span>
  <span>rescue</span> <span>ActiveRecord</span><span>::</span><span>SerializationFailure</span>
    <span>retry</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p><em>Serializable</em> is the strictest isolation level offered by PostgreSQL, but that comes with a cost. Preventing these race conditions requires managing locks and delaying or rejecting some concurrent tasks, reducing the write throughput of the database. So for many applications, it may not be desirable, and one of the other techniques above should be considered instead. In particular, you should aim for the finest-grained locks that will give you the desired result, locking over particular records or values, rather than locking a whole table on every write.</p>
<p>Now, there are many other kinds of race conditions possible in web applications, but one thing that makes them easier to reason about is <em>forms</em>. One nice thing about forms in terms of relating user expectations to internal semantics is that they&rsquo;re <em>complete</em>: when you submit a form, all its inputs and their values are sent to the server, not just the ones you&rsquo;ve changed. You&rsquo;re submitting a complete representation of the resource to the server, which can validate and accept it as such. If you submit a form on a website, and the site indicates this was successful, your expectation is that <em>all</em> the values you saw in the form are what was saved, not some combination of your changes and someone else&rsquo;s.</p>
<p>However, it is still possible to end up with a hybrid state under concurrent requests, even an invalid state, because of a performance optimisation in how Rails updates records. An <code>update</code> action in a Rails controller typically does something like this, interspersed with access control, validation, generating the response and so on:</p>
<div>
<div>
<pre><code>  <span>def</span> <span>update</span>
    <span>thing</span> <span>=</span> <span>Thing</span><span>.</span><span>find</span><span>(</span><span>params</span><span>[</span><span>:id</span><span>])</span>
    <span>thing</span><span>.</span><span>update</span><span>(</span><span>params</span><span>[</span><span>:thing</span><span>])</span>
  <span>end</span>
</code></pre>
</div>
</div>
<p>What happens here is that state from the database and the incoming <code>POST</code> request is merged, as follows. First the <code>id</code> URL parameter is used to <code>SELECT</code> a record. This makes a copy of the database record in memory, as a <code>Thing</code> instance. Then, the <code>update</code> method does three things. First, the attributes of this in-memory object are updated from the form parameters stored in <code>params[:thing]</code>. Second, the model&rsquo;s validations are run against the in-memory attributes. If this is successful, then an <code>UPDATE</code> query is sent to the database to store the in-memory state into the database.</p>
<p>As a performance optimisation, Rails does not include the complete object state in the <code>UPDATE</code> query. It only includes the fields that have changed, i.e. those that differ between the values obtained from the initial <code>SELECT</code> and those provided in the form data <code>params[:thing]</code>. This creates the possibility that some other process modifies the same record between your <code>SELECT</code> and <code>UPDATE</code>, altering fields you didn&rsquo;t change, but in a way that would make the record invalid, even though each change is valid on its own &ndash; another form of write skew.</p>
<p>Let&rsquo;s say we have the following schema for storing events, which have a <code>name</code>, and two dates: <code>starts_on</code> and <code>ends_on</code>.</p>
<div>
<div>
<pre><code><span>ActiveRecord</span><span>::</span><span>Schema</span><span>.</span><span>define</span> <span>do</span>
  <span>create_table</span> <span>:events</span><span>,</span> <span>force: </span><span>true</span> <span>do</span> <span>|</span><span>t</span><span>|</span>
    <span>t</span><span>.</span><span>string</span> <span>:name</span><span>,</span> <span>null: </span><span>false</span>
    <span>t</span><span>.</span><span>date</span> <span>:starts_on</span><span>,</span> <span>null: </span><span>false</span>
    <span>t</span><span>.</span><span>date</span> <span>:ends_on</span><span>,</span> <span>null: </span><span>false</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>In the <code>Event</code> model that represents this table, we&rsquo;ll implement a validation rule that the <code>ends_on</code> date must always be greater than or equal to the <code>starts_on</code> date, that is an event does not finish before it starts.</p>
<div>
<div>
<pre><code><span>class</span> <span>Event</span> <span>&lt;</span> <span>ActiveRecord</span><span>::</span><span>Base</span>
  <span>validate</span> <span>do</span>
    <span>unless</span> <span>starts_on</span> <span>&lt;=</span> <span>ends_on</span>
      <span>errors</span><span>.</span><span>add</span><span>(</span><span>:base</span><span>,</span> <span>"Events must start before they end"</span><span>)</span>
    <span>end</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>Our expectation here is that Rails should make sure that all objects written to the database are valid, because it runs the record&rsquo;s validations as a precondition for sending an <code>UPDATE</code> query. We&rsquo;ll run an experiment that tests this, by running the following scenario a hundred times. First, we call <code>Event.create</code> to create an event that begins tomorrow and ends in four days&rsquo; time, so it lasts for three days. We then start two threads; one changes the start date of the event and one changes the end date. After a hundred iterations, we print the total number of events, and the number that are valid according to the above logic.</p>
<div>
<div>
<pre><code><span>100</span><span>.</span><span>times</span> <span>do</span> <span>|</span><span>i</span><span>|</span>
  <span>dates</span> <span>=</span> <span>{</span> <span>starts_on: </span><span>1</span><span>.</span><span>day</span><span>.</span><span>from_now</span><span>,</span> <span>ends_on: </span><span>4</span><span>.</span><span>days</span><span>.</span><span>from_now</span> <span>}</span>
  <span>Event</span><span>.</span><span>create</span><span>(</span><span>name: </span><span>"event-</span><span>#{</span><span>i</span><span>}</span><span>"</span><span>,</span> <span>**</span><span>dates</span><span>)</span>

  <span>threads</span> <span>=</span> <span>[</span>
    <span>Thread</span><span>.</span><span>new</span> <span>{</span> <span>move_start_date</span><span>(</span><span>"event-</span><span>#{</span><span>i</span><span>}</span><span>"</span><span>)</span> <span>},</span>
    <span>Thread</span><span>.</span><span>new</span> <span>{</span> <span>move_end_date</span><span>(</span><span>"event-</span><span>#{</span><span>i</span><span>}</span><span>"</span><span>)</span> <span>}</span>
  <span>]</span>

  <span>threads</span><span>.</span><span>each</span><span>(</span><span>&amp;</span><span>:join</span><span>)</span>
<span>end</span>

<span>p</span> <span>[</span><span>:count</span><span>,</span> <span>Event</span><span>.</span><span>count</span><span>]</span>
<span>p</span> <span>[</span><span>:valid</span><span>,</span> <span>Event</span><span>.</span><span>count</span><span>(</span><span>&amp;</span><span>:valid?</span><span>)]</span>
</code></pre>
</div>
</div>
<p>Our initial implementations for <code>move_start_date</code> and <code>move_end_date</code> will be to load the relevant record using <code>Event.find</code>, and then call <code>Event#update</code>. <code>move_start_date</code> increments the <code>starts_on</code> date by two days, and <code>move_end_date</code> decrements the <code>ends_on</code> date by the same amount. Each of these changes is valid on its own, but applying both to the same record would violate the validation rule we defined above.</p>
<div>
<div>
<pre><code><span>def</span> <span>move_start_date</span><span>(</span><span>name</span><span>)</span>
  <span>evt</span> <span>=</span> <span>Event</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>name</span><span>)</span>
  <span>evt</span><span>.</span><span>update</span><span>(</span><span>starts_on: </span><span>evt</span><span>.</span><span>starts_on</span> <span>+</span> <span>2</span><span>.</span><span>days</span><span>,</span> <span>ends_on: </span><span>evt</span><span>.</span><span>ends_on</span><span>)</span>
<span>end</span>

<span>def</span> <span>move_end_date</span><span>(</span><span>name</span><span>)</span>
  <span>evt</span> <span>=</span> <span>Event</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>name</span><span>)</span>
  <span>evt</span><span>.</span><span>update</span><span>(</span><span>starts_on: </span><span>evt</span><span>.</span><span>starts_on</span><span>,</span> <span>ends_on: </span><span>evt</span><span>.</span><span>ends_on</span> <span>-</span> <span>2</span><span>.</span><span>days</span><span>)</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>To simulate how <code>update</code> is usually called in web applications with forms, I am passing both <code>starts_on</code> and <code>ends_on</code> to the <code>update</code> call, even though only one attribute is being changed in each case. Forms always send all their inputs&rsquo; values, and those values are passed into the model, so I want to make it clear that the following bug is not a result of me passing a subset of the attributes to <code>update</code>.</p>
<p>These two functions are executed concurrently, and each one performs a <code>SELECT</code> to load the record with the given <code>name</code>, and then an <code>UPDATE</code> against the same record. Let&rsquo;s say the initial insert is:</p>
<div>
<div>
<pre><code><span>INSERT</span> <span>INTO</span> <span>events</span> <span>(</span><span>name</span><span>,</span> <span>starts_on</span><span>,</span> <span>ends_on</span><span>)</span>
<span>VALUES</span> <span>(</span><span>'event-42'</span><span>,</span> <span>2020</span><span>-</span><span>09</span><span>-</span><span>01</span><span>,</span> <span>2020</span><span>-</span><span>09</span><span>-</span><span>04</span><span>);</span>
</code></pre>
</div>
</div>
<p>Now, suppose <code>move_start_date</code> runs first. It runs a <code>SELECT</code> query and will get the <code>Event</code> in its initial state, with <code>starts_on</code> and <code>ends_on</code> three days apart. It will then modify <code>starts_on</code> so that this in-memory copy of the record has a duration of one day. Rails will validate this object, see that <code>starts_on &lt;= ends_on</code> as required, and therefore send an <code>UPDATE</code> query to persist this state to the database. When <code>move_end_date</code> runs, its <code>SELECT</code> will see the <code>Event</code> either before or after the <code>UPDATE</code> from <code>move_start_date</code> has been committed. We&rsquo;d then expect that one of the following happens:</p>
<ol>
<li>If the <code>Event</code> includes the change from <code>move_start_date</code>, then changing <code>ends_on</code> will result in an invalid object where <code>starts_on &gt; ends_on</code>, and Rails should not save it to the database.</li>
<li>If the <code>Event</code> is in its initial state, then changing its <code>ends_on</code> date will result in a valid object, and Rails will persist that object state to the database.</li>
</ol>
<p>Either way, we expect the database record to remain valid and obey the rules encoded in the <code>Event</code> model. But, that&rsquo;s not what happens, because Rails generates <code>UPDATE</code> queries that only include fields that have been changed in memory, relative to the state retrieved in the <code>SELECT</code> query. In case 2, that means it sends the following queries:</p>
<div>
<div>
<pre><code><span>SELECT</span> <span>*</span> <span>FROM</span> <span>events</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'event-42'</span><span>;</span>
<span>--&gt; { id = 42, starts_on = 2020-09-01, ends_on = 2020-09-04 }</span>

<span>BEGIN</span><span>;</span>
<span>UPDATE</span> <span>events</span> <span>SET</span> <span>ends_on</span> <span>=</span> <span>2020</span><span>-</span><span>09</span><span>-</span><span>02</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>42</span><span>;</span>
<span>END</span><span>;</span>
</code></pre>
</div>
</div>
<p>Rails determines that the copy of the <code>Event</code> record in memory is valid, but it only persists <em>part</em> of that object to the database &ndash; the part that&rsquo;s changed since the initial <code>SELECT</code>. This allows the following sequence of events to happen:</p>
<div>
<div>
<pre><code><span>move_start_date</span>                 <span>|</span> <span>move_end_date</span>
<span>--------------------------------+--------------------------------</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>events</span> <span>...</span>        <span>|</span>
                                <span>|</span>
<span>BEGIN</span>                           <span>|</span>
                                <span>|</span> <span>SELECT</span> <span>*</span> <span>FROM</span> <span>events</span> <span>...</span>
<span>UPDATE</span> <span>events</span>                   <span>|</span>
    <span>SET</span> <span>starts_on</span> <span>=</span> <span>2020</span><span>-</span><span>09</span><span>-</span><span>03</span>  <span>|</span>
    <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span>                <span>|</span>
                                <span>|</span> <span>BEGIN</span>
<span>COMMIT</span>                          <span>|</span>
                                <span>|</span> <span>UPDATE</span> <span>events</span>
                                <span>|</span>     <span>SET</span> <span>ends_on</span> <span>=</span> <span>2020</span><span>-</span><span>09</span><span>-</span><span>02</span>
                                <span>|</span>     <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span>
                                <span>|</span>
                                <span>|</span> <span>COMMIT</span>
</code></pre>
</div>
</div>
<p>This is a form of write skew: two tasks read a record&rsquo;s current state, and based on that they modify different fields of it. Each of these changes would be valid on its own, but they are not valid when combined, and one of them would not have been made if these tasks executed sequentially and all executions followed case 1 above. When I run this experiment on PostgreSQL, 70 to 80 records out of 100 end up in an invalid state.</p>
<p>This bug is the result of an optimisation in Rails &ndash; even though web forms submit a complete representation of an object, and all that data is sent through to the model, the implementation of <code>update</code> means a record can end up with a mix of two requests&rsquo; submitted values for the record, and no program or person has verified that the combined state is valid.</p>
<p>This race condition will happen if the <code>SELECT</code> in <code>move_end_date</code> happens at any time before the <code>COMMIT</code> in <code>move_start_date</code>, so that both tasks see the record&rsquo;s initial state. In PostgreSQL, we can run these tasks inside <em>repeatable read</em> transactions, which means one of the <code>UPDATES</code> will be rejected if the transactions are concurrent &ndash; PostgreSQL treats this as a lost update since both tasks modify the same record, and indeed it would be a lost update if one request simply overwrote the other. MySQL does not provide such protection in <em>repeatable read</em>, and in <em>serializable</em> mode these tasks will deadlock.</p>
<p>The best portable protection against this bug is to use compare-and-swap, for example by taking advantage of the built-in Rails behaviour when we add a <code>lock_version</code> to the table:</p>
<div>
<div>
<pre><code><span>ActiveRecord</span><span>::</span><span>Schema</span><span>.</span><span>define</span> <span>do</span>
  <span>change_table</span> <span>:events</span> <span>do</span> <span>|</span><span>t</span><span>|</span>
    <span>t</span><span>.</span><span>integer</span> <span>:lock_version</span><span>,</span> <span>null: </span><span>false</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>We also need to catch the resulting <code>StaleObjectError</code> in our functions. I&rsquo;m handling this error by doing nothing, just to prevent invalid records ending up in the database, but you&rsquo;d probably want ask the user to resolve the conflict here rather than silently losing their input.</p>
<div>
<div>
<pre><code><span>def</span> <span>move_start_date</span><span>(</span><span>name</span><span>)</span>
  <span>evt</span> <span>=</span> <span>Event</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>name</span><span>)</span>
  <span>evt</span><span>.</span><span>update</span><span>(</span><span>starts_on: </span><span>evt</span><span>.</span><span>starts_on</span> <span>+</span> <span>2</span><span>.</span><span>days</span><span>,</span> <span>ends_on: </span><span>evt</span><span>.</span><span>ends_on</span><span>)</span>
<span>rescue</span> <span>ActiveRecord</span><span>::</span><span>StaleObjectError</span>
<span>end</span>

<span>def</span> <span>move_end_date</span><span>(</span><span>name</span><span>)</span>
  <span>evt</span> <span>=</span> <span>Event</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>name</span><span>)</span>
  <span>evt</span><span>.</span><span>update</span><span>(</span><span>starts_on: </span><span>evt</span><span>.</span><span>starts_on</span><span>,</span> <span>ends_on: </span><span>evt</span><span>.</span><span>ends_on</span> <span>-</span> <span>2</span><span>.</span><span>days</span><span>)</span>
<span>rescue</span> <span>ActiveRecord</span><span>::</span><span>StaleObjectError</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>Again, we&rsquo;ve seen that simply making <code>write()</code> fallible, changing its type from <code>write(T)</code> to <code>write(T) -&gt; bool</code> (or something more complex), can prevent race conditions that our web framework is introducing. &ldquo;Fallible&rdquo; can mean that the database outright rejects a write due to isolation rules, or it can report that the write affected no rows in a compare-and-swap, but either way it&rsquo;s not simply allowing the write to unconditionally succeed as the empty return type of <code>write(T)</code> implies.</p>
<p>In this article, we&rsquo;ve added another category of race condition to our understanding. Whereas a <em>lost update</em> means one write replacing another, a <em>write skew</em> means two writes being persisted, where one of them would have been different &ndash; or not performed at all &ndash; if the tasks were executed sequentially, often leaving your data in an inconsistent state. Everything we&rsquo;ve looked at so far is a <em>single-request</em> bug, which happens if two <code>POST</code> requests arrive at the same time. In the next instalment, we&rsquo;ll extend this concept and look at <em>multi-request</em> scenarios where the <code>Location</code> type appears in the workflow of viewing and updating resources in applications.</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via<br>
<br>
June 14, 2022 at 03:27PM
